src/Rector/AbstractRector.php

```diff
@@ -251,58 +248,40 @@ final public function enterNode(Node $node)
         /** @var Node $node */
         $this->mirrorAttributes($originalAttributes, $node);
 
-        $currentScope = $originalNode->getAttribute(AttributeKey::SCOPE);
-
-        $requiresScopeRefresh = true;
-
-        // names do not have scope in PHPStan
-        if (! $node instanceof Name && ! $node instanceof Namespace_ && ! $node instanceof FileWithoutNamespace && ! $node instanceof Identifier) {
-            if ($currentScope === null) {
-                $parent = $node->getAttribute(AttributeKey::PARENT_NODE);
-
-                // in case of unreachable stmts, no other node will have available scope
-                // loop all previous expressions, until we find nothing or is_unreachable
-                $currentStmt = $this->betterNodeFinder->resolveCurrentStatement($parent);
-
-                if ($currentStmt instanceof Stmt && $this->unreachableStmtAnalyzer->isStmtPHPStanUnreachable(
-                    $currentStmt
-                )) {
-                    $requiresScopeRefresh = false;
-                }
-
-                if ($requiresScopeRefresh) {
-                    $errorMessage = sprintf(
-                        'Node "%s" with parent of "%s" is missing scope required for scope refresh.',
-                        $node::class,
-                        $parent instanceof Node ? $parent::class : null
-                    );
-
-                    throw new ShouldNotHappenException($errorMessage);
-                }
-            }
-
-            if ($requiresScopeRefresh) {
-                $this->changedNodeScopeRefresher->refresh($node, $this->file->getSmartFileInfo(), $currentScope);
-            }
+        if (in_array(
+            $node::class,
+            [Name::class, FullyQualified::class, Namespace_::class, FileWithoutNamespace::class, Identifier::class],
+            true
+        )) {
+            return $this->processResultSingleNode($node, $originalNode);
         }
 
-        $this->connectParentNodes($node);
+        $currentScope = $originalNode->getAttribute(AttributeKey::SCOPE);
 
-        // is equals node type? return node early
-        if ($originalNode::class === $node::class) {
-            return $node;
+        if ($currentScope instanceof MutatingScope) {
+            $this->changedNodeScopeRefresher->refresh($node, $this->file->getSmartFileInfo(), $currentScope);
+            return $this->processResultSingleNode($node, $originalNode);
         }
 
-        // search "infinite recursion" in https://github.com/nikic/PHP-Parser/blob/master/doc/component/Walking_the_AST.markdown
-        $originalNodeHash = spl_object_hash($originalNode);
+        $isUnreachable = $node->getAttribute(AttributeKey::IS_UNREACHABLE);
+        if ($isUnreachable === true) {
+            return $this->processResultSingleNode($node, $originalNode);
+        }
```

Comment:
The unreachable logic should be kept outside the service, as we'll use it in another places too.
Also the `AbstractRector` is pretty huge now, so the aim is to make it as thin as possible.

Reply:
ok, re-add `UnreachableStmtAnalyzer`, with look by current stmt, directly check IS_UNREACHABLE, fallback to false when current stmt is not Stmt.